discovery/tailscale/tailscale.go

```diff
@@ -0,0 +1,187 @@
+package tailscale
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"net/url"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/common/config"
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/prometheus/discovery"
+	"github.com/prometheus/prometheus/discovery/refresh"
+	"github.com/prometheus/prometheus/discovery/targetgroup"
+
+	"tailscale.com/client/tailscale"
+)
+
+const (
+	tsLabel                                 = model.MetaLabelPrefix + "tailscale_"
+	deviceLabel                             = tsLabel + "device_"
+	tsLabelDeviceAddresses                  = deviceLabel + "addresses"
+	tsLabelDeviceID                         = deviceLabel + "id"
+	tsLabelDeviceName                       = deviceLabel + "name"
+	tsLabelDeviceHostname                   = deviceLabel + "hostname"
+	tsLabelDeviceUser                       = deviceLabel + "user"
+	tsLabelDeviceClientVersion              = deviceLabel + "client_version"
+	tsLabelDeviceUpdate                     = deviceLabel + "update"
+	tsLabelDeviceOS                         = deviceLabel + "os"
+	tsLabelDeviceKeyExpiryDisavle           = deviceLabel + "key_expiry_disabled"
+	tsLabelDeviceAuthorized                 = deviceLabel + "authorized"
+	tsLabelDeviceIsExternal                 = deviceLabel + "is_external"
+	tsLabelDeviceMachineKey                 = deviceLabel + "machine_key"
+	tsLabelDeviceNodeKey                    = deviceLabel + "node_key"
+	tsLabelDeviceBlocksIncommingConnections = deviceLabel + "blocks_incomming_connections"
+	tsLabelDeviceEnabledRoutes              = deviceLabel + "enabled_routes"
+	tsLabelDeviceAdvertisedRoutes           = deviceLabel + "advertised_routes"
+	tsLabelDeviceTags                       = deviceLabel + "tags"
+	tsLabelDeviceTailnetLockKey             = deviceLabel + "tailnet_lock_key"
+)
+
+var (
+	DefaultSDConfig = SDConfig{
+		RefreshInterval:  model.Duration(60 * time.Second),
+		URL:              "https://api.tailscale.com",
+		Tailnet:          "-",
+		TagSeparator:     ",",
+	}
+	failuresCount = prometheus.NewCounter(
+		prometheus.CounterOpts{
+			Name: "prometheus_sd_tailscale_failures_total",
+			Help: "Number of tailscale service discovery refresh failures.",
+		})
+)
+
+func init() {
+	tailscale.I_Acknowledge_This_API_Is_Unstable = true
```

Comment:
By migrating to the Tailscale SDK, this is a requirement to use the API. I'm somewhat hesitant to introduce this "unstable" in Prometheus. Then again, if this is unstable, so would the API at the server side.

Reply:
Yeah, I think that's fine. Let's just hope they don't change their API too much in the future in any case :)